### PR TITLE
Amazon MQ throughput-optimized message brokers

### LINKED ISSUE #11325: Amazon MQ throughput-optimized message brokers
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Amazon MQ supports throughput-optimized message brokers, backed by Amazon Elastic Block Store.

### New or Affected Resource(s)

* [`aws_mq_broker`](https://www.terraform.io/docs/providers/aws/r/mq_broker.html)

### Potential Terraform Configuration

```hcl
resource "aws_mq_broker" "example" {
  storage_type = "EBS"
}
```

### References

[Announcement](https://aws.amazon.com/about-aws/whats-new/2019/12/amazon-mq-introduces-throughput-optimized-message-brokers/).
[Developer guide](https://docs.aws.amazon.com/amazon-mq/latest/developer-guide/broker-storage.html).

Requires [AWS SDK v1.26.3](https://github.com/aws/aws-sdk-go/releases/tag/v1.26.3):
* https://github.com/terraform-providers/terraform-provider-aws/pull/11314

### PR DESCRIPTION
**Pull request by gdepuille**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #11325 

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
ENHANCEMENTS:

* resource/aws_mq_broker: Add possibility to create an AmazonMQ broker with EBS (throughput optimized)
```

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSMqBroker'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSMqBroker -timeout 120m
=== RUN   TestAccAWSMqBroker_basic
=== PAUSE TestAccAWSMqBroker_basic
=== RUN   TestAccAWSMqBroker_throughputOptimized
=== PAUSE TestAccAWSMqBroker_throughputOptimized
=== RUN   TestAccAWSMqBroker_allFieldsDefaultVpc
=== PAUSE TestAccAWSMqBroker_allFieldsDefaultVpc
=== RUN   TestAccAWSMqBroker_allFieldsCustomVpc
=== PAUSE TestAccAWSMqBroker_allFieldsCustomVpc
=== RUN   TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== RUN   TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
=== RUN   TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== RUN   TestAccAWSMqBroker_updateUsers
=== PAUSE TestAccAWSMqBroker_updateUsers
=== RUN   TestAccAWSMqBroker_updateTags
=== PAUSE TestAccAWSMqBroker_updateTags
=== RUN   TestAccAWSMqBroker_updateSecurityGroup
=== PAUSE TestAccAWSMqBroker_updateSecurityGroup
=== CONT  TestAccAWSMqBroker_basic
=== CONT  TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== CONT  TestAccAWSMqBroker_updateTags
=== CONT  TestAccAWSMqBroker_updateUsers
=== CONT  TestAccAWSMqBroker_updateSecurityGroup
=== CONT  TestAccAWSMqBroker_allFieldsCustomVpc
=== CONT  TestAccAWSMqBroker_allFieldsDefaultVpc
=== CONT  TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== CONT  TestAccAWSMqBroker_throughputOptimized
=== CONT  TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
--- PASS: TestAccAWSMqBroker_throughputOptimized (795.27s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_KmsKeyId (1126.72s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled (1145.42s)
--- PASS: TestAccAWSMqBroker_updateTags (1186.27s)
--- PASS: TestAccAWSMqBroker_basic (1201.82s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled (1227.87s)
--- PASS: TestAccAWSMqBroker_updateUsers (1567.22s)
--- PASS: TestAccAWSMqBroker_updateSecurityGroup (1573.84s)
--- PASS: TestAccAWSMqBroker_allFieldsCustomVpc (1810.47s)
--- PASS: TestAccAWSMqBroker_allFieldsDefaultVpc (2039.51s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	2039.562s

```

### CODE REVIEW

**Comment by gdepuille** [comment]
Hi @bflad, 
Can you communicate an ETA to validate this PR ?

**Comment by gdepuille** [comment]
Hi,

I've rebased this development over the last release.

I've relaunched all acceptance tests and all are passed on the MQ features.

```bash
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSMq'

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSMq -timeout 120m
=== RUN   TestAccAWSMqBroker_basic
=== PAUSE TestAccAWSMqBroker_basic
=== RUN   TestAccAWSMqBroker_throughputOptimized
=== PAUSE TestAccAWSMqBroker_throughputOptimized
=== RUN   TestAccAWSMqBroker_allFieldsDefaultVpc
=== PAUSE TestAccAWSMqBroker_allFieldsDefaultVpc
=== RUN   TestAccAWSMqBroker_allFieldsCustomVpc
=== PAUSE TestAccAWSMqBroker_allFieldsCustomVpc
=== RUN   TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== RUN   TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
=== RUN   TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== RUN   TestAccAWSMqBroker_updateUsers
=== PAUSE TestAccAWSMqBroker_updateUsers
=== RUN   TestAccAWSMqBroker_updateTags
=== PAUSE TestAccAWSMqBroker_updateTags
=== RUN   TestAccAWSMqBroker_updateSecurityGroup
=== PAUSE TestAccAWSMqBroker_updateSecurityGroup
=== RUN   TestAccAWSMqBroker_disappears
=== PAUSE TestAccAWSMqBroker_disappears
=== RUN   TestAccAWSMqConfiguration_basic
=== PAUSE TestAccAWSMqConfiguration_basic
=== RUN   TestAccAWSMqConfiguration_withData
=== PAUSE TestAccAWSMqConfiguration_withData
=== RUN   TestAccAWSMqConfiguration_updateTags
=== PAUSE TestAccAWSMqConfiguration_updateTags
=== CONT  TestAccAWSMqBroker_basic
=== CONT  TestAccAWSMqBroker_updateTags
=== CONT  TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== CONT  TestAccAWSMqBroker_updateUsers
=== CONT  TestAccAWSMqConfiguration_withData
=== CONT  TestAccAWSMqConfiguration_updateTags
=== CONT  TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== CONT  TestAccAWSMqBroker_allFieldsDefaultVpc
=== CONT  TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
=== CONT  TestAccAWSMqBroker_disappears
=== CONT  TestAccAWSMqBroker_allFieldsCustomVpc
=== CONT  TestAccAWSMqConfiguration_basic
=== CONT  TestAccAWSMqBroker_throughputOptimized
=== CONT  TestAccAWSMqBroker_updateSecurityGroup
--- PASS: TestAccAWSMqConfiguration_withData (36.82s)
--- PASS: TestAccAWSMqConfiguration_basic (61.86s)
--- PASS: TestAccAWSMqConfiguration_updateTags (89.84s)
--- PASS: TestAccAWSMqBroker_throughputOptimized (852.44s)
--- PASS: TestAccAWSMqBroker_basic (1140.31s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled (1141.84s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_KmsKeyId (1161.45s)
--- PASS: TestAccAWSMqBroker_disappears (1170.60s)
--- PASS: TestAccAWSMqBroker_updateTags (1234.79s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled (1261.30s)
--- PASS: TestAccAWSMqBroker_updateSecurityGroup (1484.36s)
--- PASS: TestAccAWSMqBroker_updateUsers (1584.28s)
--- PASS: TestAccAWSMqBroker_allFieldsDefaultVpc (2005.72s)
--- PASS: TestAccAWSMqBroker_allFieldsCustomVpc (2012.93s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	2013.016s

```

**Comment by gdepuille** [comment]
I've realized a rebase from 3.7.0

**Comment by matosf** [comment]
Any ETA for this to be merged?

**Comment by ziek-za** [comment]
@gdepuille are there any updates? Would love to be able to make use of this!

**Comment by gdepuille** [comment]
@ziek-za I'm sorry that this PR is not included in the basic version. I use this in production since it's submitted. If you want to use it, it's possible to make a custom version of the provider with the commits you need. It's just you need some piece of time when new features are added in the core version.

**Comment by YakDriver** [comment]
@gdepuille Thank you for this PR! We performed significant MQ reorganization in #16108. It looks like your PR may have some helpful tests beyond that. Please rebase when you have a chance.

**Comment by gdepuille** [comment]
@YakDriver I've just been rebase this one. I've relaunch acceptance tests

<details>
  <summary><pre>PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1772.201s</pre></summary>

<pre>$> make testacc TEST=./aws TESTARGS=&apos;-run=TestAccAWSMq&apos;==&gt; Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSMq -timeout 120m
=== RUN   TestAccAWSMqBroker_basic
=== PAUSE TestAccAWSMqBroker_basic
=== RUN   TestAccAWSMqBroker_throughputOptimized
=== PAUSE TestAccAWSMqBroker_throughputOptimized
=== RUN   TestAccAWSMqBroker_allFieldsDefaultVpc
=== PAUSE TestAccAWSMqBroker_allFieldsDefaultVpc
=== RUN   TestAccAWSMqBroker_allFieldsCustomVpc
=== PAUSE TestAccAWSMqBroker_allFieldsCustomVpc
=== RUN   TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== RUN   TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
=== RUN   TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== PAUSE TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== RUN   TestAccAWSMqBroker_updateUsers
=== PAUSE TestAccAWSMqBroker_updateUsers
=== RUN   TestAccAWSMqBroker_updateTags
=== PAUSE TestAccAWSMqBroker_updateTags
=== RUN   TestAccAWSMqBroker_updateSecurityGroup
=== PAUSE TestAccAWSMqBroker_updateSecurityGroup
=== RUN   TestAccAWSMqBroker_disappears
=== PAUSE TestAccAWSMqBroker_disappears
=== RUN   TestAccAWSMqBroker_rabbitmq
=== PAUSE TestAccAWSMqBroker_rabbitmq
=== RUN   TestAccAWSMqConfiguration_basic
=== PAUSE TestAccAWSMqConfiguration_basic
=== RUN   TestAccAWSMqConfiguration_withData
=== PAUSE TestAccAWSMqConfiguration_withData
=== RUN   TestAccAWSMqConfiguration_updateTags
=== PAUSE TestAccAWSMqConfiguration_updateTags
=== CONT  TestAccAWSMqBroker_basic
=== CONT  TestAccAWSMqBroker_rabbitmq
=== CONT  TestAccAWSMqBroker_allFieldsDefaultVpc
=== CONT  TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled
=== CONT  TestAccAWSMqBroker_updateSecurityGroup
=== CONT  TestAccAWSMqBroker_updateTags
=== CONT  TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled
=== CONT  TestAccAWSMqBroker_EncryptionOptions_KmsKeyId
=== CONT  TestAccAWSMqConfiguration_updateTags
=== CONT  TestAccAWSMqConfiguration_withData
=== CONT  TestAccAWSMqConfiguration_basic
=== CONT  TestAccAWSMqBroker_updateUsers
=== CONT  TestAccAWSMqBroker_throughputOptimized
=== CONT  TestAccAWSMqBroker_disappears
=== CONT  TestAccAWSMqBroker_allFieldsCustomVpc
--- PASS: TestAccAWSMqConfiguration_withData (34.41s)
--- PASS: TestAccAWSMqConfiguration_basic (52.58s)
--- PASS: TestAccAWSMqConfiguration_updateTags (68.85s)
--- PASS: TestAccAWSMqBroker_rabbitmq (571.55s)
--- PASS: TestAccAWSMqBroker_throughputOptimized (861.68s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_KmsKeyId (1085.95s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled (1165.41s)
--- PASS: TestAccAWSMqBroker_disappears (1197.57s)
--- PASS: TestAccAWSMqBroker_basic (1214.57s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled (1223.27s)
--- PASS: TestAccAWSMqBroker_updateTags (1239.36s)
--- PASS: TestAccAWSMqBroker_updateSecurityGroup (1357.05s)
--- PASS: TestAccAWSMqBroker_updateUsers (1604.82s)
--- PASS: TestAccAWSMqBroker_allFieldsDefaultVpc (1710.93s)
--- PASS: TestAccAWSMqBroker_allFieldsCustomVpc (1772.14s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1772.201s
</pre>
</details>

File: website/docs/r/mq_broker.html.markdown [hunk 1]
```diff
@@ -42,14 +44,39 @@ resource "aws_mq_broker" "example" {
 }
 ```
 
+### Broker with EBS storage (high throughput optimized only on ActiveMQ engine type)
```
Comment (YakDriver) [inline website/docs/r/mq_broker.html.markdown:47]:
```suggestion
### High-throughput Optimized Example

This example shows the use of EBS storage for high-throughput optimized performance. Currently, this only works with `ActiveMQ`.
```

File: website/docs/r/mq_broker.html.markdown [hunk 2]
```diff
@@ -21,6 +21,8 @@ Provides an Amazon MQ broker resource. This resources also manages users for the
 
 ## Example Usage
 
+### Broker with EFS storage
```
Comment (YakDriver) [inline website/docs/r/mq_broker.html.markdown:24]:
```suggestion
### Basic Example
```

File: website/docs/r/mq_broker.html.markdown [hunk 3]
```diff
@@ -65,7 +92,7 @@ The following arguments are optional:
 * `maintenance_window_start_time` - (Optional) Configuration block for the maintenance window start time. Detailed below.
 * `publicly_accessible` - (Optional) Whether to enable connections from applications outside of the VPC that hosts the broker's subnets.
 * `security_groups` - (Optional) List of security group IDs assigned to the broker.
-* `storage_type` - (Optional) Storage type of the broker. For `engine_type` `ActiveMQ`, the valid values are `efs` and `ebs`. For `engine_type` `RabbitMQ`, only `EBS` is supported.
+* `storage_type` - (Optional) Storage type of the broker. For `engine_type` `ActiveMQ`, the valid values are `efs` and `ebs`. When `ebs` storage type is defined, only the `mq.m5` broker instance type family is supported. For `engine_type` `RabbitMQ`, only `EBS` is supported.
```
Comment (YakDriver) [inline website/docs/r/mq_broker.html.markdown:95]:
```suggestion
* `storage_type` - (Optional) Storage type of the broker. For `engine_type` `ActiveMQ`, the valid values are `efs` and `ebs`. When using `ebs`, only the `mq.m5` broker instance type family is supported. For `engine_type` `RabbitMQ`, only `ebs` is supported. The AWS-default is `ebs`.
```

File: aws/resource_aws_mq_broker_test.go [hunk 4]
```diff
@@ -278,6 +278,7 @@ func TestAccAWSMqBroker_basic(t *testing.T) {
 					resource.TestCheckResourceAttr(resourceName, "encryption_options.0.use_aws_owned_key", "true"),
 					resource.TestCheckResourceAttr(resourceName, "engine_type", "ActiveMQ"),
 					resource.TestCheckResourceAttr(resourceName, "engine_version", "5.15.0"),
+					resource.TestCheckResourceAttr(resourceName, "storage_type", "efs"),
```
Comment (YakDriver) [inline aws/resource_aws_mq_broker_test.go:281]:
```suggestion
```

This is a duplicate.

**Review by YakDriver — CHANGES_REQUESTED** [review]
Thanks for your changes! There are just a few minor fixes and we should be good to go!

**Comment by gdepuille** [comment]
@YakDriver I trust you for the documentation because it's not my native language ;-)

I've validated all your suggestion. I'm very happy this PR is near to be validated.

**Review by YakDriver — APPROVED** [review]
Looks good! Thanks!

Acceptance tests:
```
--- PASS: TestAccAWSMqBroker_rabbitmq (549.56s)
--- PASS: TestAccAWSMqBroker_throughputOptimized (791.92s)
--- PASS: TestAccAWSMqBroker_disappears (1138.54s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Disabled (1182.12s)
--- PASS: TestAccAWSMqBroker_basic (1191.68s)
--- PASS: TestAccAWSMqBroker_updateTags (1199.67s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_UseAwsOwnedKey_Enabled (1225.50s)
--- PASS: TestAccAWSMqBroker_EncryptionOptions_KmsKeyId (1286.37s)
--- PASS: TestAccAWSMqBroker_updateSecurityGroup (1346.46s)
--- PASS: TestAccAWSMqBroker_updateUsers (1590.61s)
--- PASS: TestAccAWSMqBroker_allFieldsDefaultVpc (1696.42s)
--- PASS: TestAccAWSMqBroker_allFieldsCustomVpc (1748.95s)
```

**Pull request merged**